src/pd/async/client.rs

```diff
@@ -0,0 +1,217 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+
+use protobuf::RepeatedField;
+
+use futures::Future;
+
+use kvproto::metapb;
+use kvproto::pdpb::{self, Member};
+use kvproto::pdpb_grpc::PDAsync;
+
+use super::super::{Result, Error, PdFuture};
+use super::super::AsyncPdClient;
+use super::validate_endpoints;
+
+use super::util::LeaderClient;
+use super::util::Request;
+
+// TODO: revoke pubs.
+pub struct RpcAsyncClient {
+    pub cluster_id: u64,
+    pub inner: LeaderClient,
+}
+
+impl RpcAsyncClient {
+    pub fn new(endpoints: &str) -> Result<RpcAsyncClient> {
+        let endpoints: Vec<_> = endpoints.split(',')
+            .map(|s| if !s.starts_with("http://") {
+                format!("http://{}", s)
+            } else {
+                s.to_owned()
+            })
+            .collect();
+
+        let (client, members) = try!(validate_endpoints(&endpoints));
+
+        Ok(RpcAsyncClient {
+            cluster_id: members.get_header().get_cluster_id(),
+            inner: LeaderClient::new(client, members),
+        })
+    }
+
+    pub fn header(&self) -> pdpb::RequestHeader {
+        let mut header = pdpb::RequestHeader::new();
+        header.set_cluster_id(self.cluster_id);
+        header
+    }
+
+    // For tests
+    pub fn get_leader(&self) -> Member {
+        self.inner.clone_members().take_leader()
+    }
+}
+
+fn check_resp_header(header: &pdpb::ResponseHeader) -> Result<()> {
+    if !header.has_error() {
+        return Ok(());
+    }
+    // TODO: translate more error types
+    let err = header.get_error();
+    Err(box_err!(err.get_message()))
+}
+
+impl fmt::Debug for RpcAsyncClient {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(fmt,
+               "PD gRPC Client connects to cluster {:?}",
```

Comment:
A plain text description is not suitable for a struct. Add type name,  and wrap all in square brackets.